### ISSUE TITLE
no-unused-vars and no-undef with modules

### ISSUE THREAD

**Issue by wavded** [eslint/eslint#1999]
Given this code:

```js
import React from 'react'
import { Route, DefaultRoute, NotFoundRoute } from 'react-router'

import App from './app'
import Index from './pages/index'
import NotFound from './pages/not-found'
import Dashboard from './pages/dashboard'

let routes = (
  <Route name="app" path="/" handler={App}>
    <DefaultRoute name="index" handler={Index} />
    <Route name="dashboard" handler={Dashboard} />
    <NotFoundRoute name="notfound" handler={NotFound} />
  </Route>
)

export default routes
```

I'm getting `no-unused-vars` and `no-undef` errors (even after escope update in https://github.com/eslint/eslint/issues/1978)

**Comment by nzakas** [comment]
Can you include the actual errors? It really helps to see what's being flagged.

**Comment by wavded** [comment]
Sure thing:

```
routes.js
   1:7   warning  React is defined but never used          no-unused-vars
   2:9   warning  Route is defined but never used          no-unused-vars
   2:16  warning  DefaultRoute is defined but never used   no-unused-vars
   2:30  warning  NotFoundRoute is defined but never used  no-unused-vars
  10:3   error    'Route' is not defined                   no-undef
  11:5   error    'DefaultRoute' is not defined            no-undef
  12:5   error    'Route' is not defined                   no-undef
  13:5   error    'NotFoundRoute' is not defined           no-undef

✖ 8 problems (4 errors, 4 warnings)
```

**Comment by wavded** [comment]
Perhaps this is a red herring as I noticed the latest `0.16.2` doesn't actually have escope `2.0.7` changes but rather the version is fixed to `2.0.6`.

**Comment by wavded** [comment]
Guess I'm not sure, even if force escope to `2.0.7` I still get the same errors.

**Comment by ivanoats** [comment]
I think I am getting the same issue here:

```
ivan  ~  dev  scholarships  eslint client/js/components/donorCard.js                                                                                                                        develop

client/js/components/donorCard.js
  3:4   error  DonorCard is defined but never used  no-unused-vars
  3:16  error  'React' is not defined               no-undef
  5:12  error  'React' is not defined               no-undef
  6:10  error  'React' is not defined               no-undef
  7:14  error  'React' is not defined               no-undef
  8:14  error  'React' is not defined               no-undef
  9:15  error  'React' is not defined               no-undef

✖ 7 problems (7 errors, 0 warnings)

 ivan  ~  dev  scholarships  ccat client/js/components/donorCard.js                                                                                                                     1   develop
```
```javascript
import React from 'react';

let DonorCard = React.createClass({
  propTypes: {
    header: React.PropTypes.string,
    body: React.PropTypes.string,
    imageUrl: React.PropTypes.string,
    imageAlt: React.PropTypes.string,
    draggable: React.PropTypes.string
  },
  render() {
    return (
      <div className="card" draggable="true">
        <h2>Donor: {this.props.header}</h2>
        <div className="card-image">
          <img src={this.props.imageUrl} alt={this.props.imageAlt} />
        </div>
        <div className="card-header">
          A header
        </div>
        <div className="card-copy">
          <p>{this.props.body}</p>
        </div>
      </div>
    );
  }
});

export { DonorCard };
```

**Comment by ai** [comment]
Seems like #1978 issue is back :(. I have same problem.

**Comment by nzakas** [comment]
Okay, JSX is a known issue and we are changing how that works. See #1911.

**Comment by ericclemmons** [comment]
@nzakas I don't think it's JSX specific:

```javascript
import router from "koa-router";

export default router()
  .get("/howdy", function *() {
    this.body = 'there';
  })
;
```

```
  3:15  error  'router' is not defined  no-undef
```

```
    "eslint": "^0.16.2",
```

**Comment by nzakas** [comment]
There are known issues with modules due to escope issues. Please look at open issues before commenting on closed ones.

**Comment by ericclemmons** [comment]
@nzakas I've been following them for a few days & the open ones are closed (as saying fixed), so it's like whack-a-mole up in here! :D

**Comment by nzakas** [comment]
#1978 

It is whack a mole. Being the first linter with ES6 support means we find all the bugs. :)

**Comment by ericclemmons** [comment]
_AND_ ES7 :D

**Subscribed to #1978!**

Thanks!
